functorch/csrc/BatchRulesScatterOps.cpp

```diff
@@ -102,25 +155,152 @@ std::tuple<Tensor,optional<int64_t>> index_batch_rule(
+      // indices: [:, Tensor[2, 2], Tensor[2, 2], :]
+      // batched_indices: [:, :, Tensor[2, 2], Tensor[2, 2], :]
+      // res: Tensor[B, 5, 2, 2, 8]
+      return std::make_tuple(res, 0);
+    } else {
+      // self: Tensor[B, 5, 6, 7]
+      // indices: [Tensor[2, 2], :, Tensor[2, 2]]
+      // batched_indices: [:, Tensor[2, 2], :, Tensor[2, 2]]
+      // res: Tensor[2, 2, B, 6]
+      return std::make_tuple(res, max_index_dim);
+    }
+  }
+
+  // Case 2
+  if (!self_batched && indices_batched) {
+    if (advanced_indices_are_adjacent) {
+      // self: Tensor[5, 6, 7, 8]
+      // indices: [:, :, Tensor[B, 2, 2], Tensor[2, 2]]
+      // batched_indices: indices (no change)
+      // res: Tensor[5, 6, B, 2, 2]
+      return std::make_tuple(res, num_leading_nones);
+    } else {
+      // self: Tensor[5, 6, 7, 8, 9]
+      // indices: [:, :, Tensor[B, 2, 2], :, Tensor[2, 2]]
+      // batched_indices: indices (no change)
+      // res: Tensor[B, 2, 2, 5, 6, 8]
+      return std::make_tuple(res, 0);
     }
-    outDim = 0;
   }
-  return std::make_tuple(res, outDim);
+
+  // Case 3: self_batched and indices_batched
+  TORCH_INTERNAL_ASSERT(self_batched && indices_batched);
+  if (!advanced_indices_are_adjacent) {
+    // self: Tensor[B, 5, 6, 7, 8]
+    // indices: [:, Tensor[B, 2, 2], :, Tensor[2, 2]]
+    // batched_indices: [arange(B).expand(B, 2, 2), :, Tensor[B, 2, 2], :, Tensor[2, 2]]
+    // res: Tensor[B, 2, 2, 5, 7]
+    return std::make_tuple(res, 0);
+  }
+  // In other words, in batched_indices, advanced indices are adjacent
+  if (num_leading_nones == 0) {
+    // self: Tensor[B, 5, 6, 7, 8]
+    // indices: [Tensor[B, 2, 2], Tensor[2, 2], :, :]
+    // batched_indices: [arange(B).expand(B, 2, 2), Tensor[B, 2, 2], Tensor[2, 2], :, :]
+    // res: Tensor[B, 2, 2, 7, 8]
+    return std::make_tuple(res, 0);
+  }
+  // This is the tricky case. In indices, advanced indices are adjacent.
+  // In batched_indices, advanced indices are no longer adjacent
+  //
+  // self: Tensor[B, 5, 6, 7, 8, 9]
+  // indices: [:, :, Tensor[B, 2, 3], Tensor[2, 3], :]
+  // batched_indices: [arange(B).expand(B, 2, 3), :, :, Tensor[B, 2, 3], Tensor[2, 3], :]
+  // res: Tensor[B, 2, 3, 5, 6, 9]
+  // expected: Tensor[B, 5, 6, 2, 3, 9]
+  //
+  // The resolution is to move dims around until we get the right shape.
+  // The result is set up as [B, <maxIndexDim>, <leading_nones>, ...]
+  // we just have to move the <leading_nones> to before the <maxIndexDim> to produce
+  // [B, <leading_nones>, <maxIndexDim>, ...]
+  VmapDimVector permutation(res.dim(), 0);
+  std::iota(permutation.begin(), permutation.end(), 0);
+  std::iota(
+      permutation.begin() + 1,
+      permutation.begin() + 1 + num_leading_nones,
+      1 + max_index_dim);
+  std::iota(
+      permutation.begin() + 1 + num_leading_nones,
+      permutation.begin() + 1 + max_index_dim + num_leading_nones,
+      1);
+  return std::make_tuple(res.permute(permutation), 0);
```

Comment:
nit: since we're doing the same thing for this here and in maybe_permute values and it has such a thorough explanation, it might be nice to have this as a helper function

Reply:
Good idea